### DISCUSSION TITLE
Constructing and retrieving RM3

### DISCUSSION THREAD

**Discussion by eyalelani** [castorini/pyserini#1667] [Q&A]
Hi, I would like to know if there is an option of retrieving the RM3 model, which is based on some list of documents.
As far as I know, in Pyserini the usage of RM3 is only for reranking documents that were firstly ranked by BM25, and then the output is the list of the documents after they are being reranked.

However, given a list of documents, I would like to get **the RM3 model** based on these documents. That is, as an output I would like to have something like a  dictionary of terms and their probabilities in the created model.

I've noticed that in Anserini's implementation there exists a private function that returns a model, based on scored documents (exactly what I need),  whose signature is:
private FeatureVector estimateRelevanceModel(ScoredDocuments docs, IndexReader reader, boolean tweetsearch, boolean useRf);
The problem is that it's written in Java and I don't know how to use it in Python. I'm aware of the option to use pyserini.pyclass.autoclass, though I'm not sure if I can use it, and if I do - how.

If it may help, my goal is using this model with **another** index that I've created, of the documents' passages.

I'll be glad to have any kind of help,
Eyal

**Comment by octocat20** [comment]
You are right that the usual Pyserini RM3 path is BM25 hits, then feedback, then a new query. The CLI does not hand you the term probability dict.

If you already have a doc list and want the RM3 model itself, use the Python rerank helpers:

- get vectors with LuceneIndexReader.get_document_vector(docid)
- call RM3Reranker.compute_relevance_model(vectors, document_scores, fb_terms=...)

That gives you a normalized term -> weight map. From there you can interpolate with the original query or apply those terms on your passage index.

You do not need Anserini's private estimateRelevanceModel for this. compute_relevance_model is the Python version of "docs in, relevance model out".
